In [22]:
import pandas as pd
from pathlib import Path

# Load all association partitions
associations_dir = Path('../data/raw/direct_associations')
association_files = sorted(associations_dir.glob('*.parquet'))

# Read and concatenate all partitions
a_df = pd.concat([pd.read_parquet(f) for f in association_files], ignore_index=True)

# Load the biomarkers file
b_df = pd.read_parquet("../data/raw/evidence_cancer_biomarkers/evidence_cancer_biomarkers.parquet")

In [31]:
print(f"Loaded {len(association_files)} partitions")
print(f"Total rows: {a_df.shape[0]:,}")

a_df.head()

Loaded 20 partitions
Total rows: 4,492,971


diseaseId         targetId     score  evidenceCount
0  DOID_0050890  ENSG00000001084  0.031799              4
1  DOID_0050890  ENSG00000002549  0.001478              1
2  DOID_0050890  ENSG00000004142  0.002217              1
3  DOID_0050890  ENSG00000004478  0.002217              1
4  DOID_0050890  ENSG00000004948  0.002957              1

In [40]:
b_df.head(50)

id targetFromSourceId  \
0   0010fb8e532bdb5d0072cc5db9991923f3c4b121               EGFR   
1   0023bc1ee01a34cf68f6e8d55e0b4466893114b8            SMARCB1   
2   006c16ffc7eb3dc04bb2ccf3ce4994bb1990fc9d               EGFR   
3   007f807da0c31088ec4a9136812edb6f70d87701                NF1   
4   00b3e7749de158868d86a20a241ef3122f932137              ERCC1   
5   0119fae7b3f606cce180328d0ad893706586429a             MAP2K1   
6   01a588cbcd96623bb612b85abfc97e51b93bbaa0               EGFR   
7   01a6a63ad25808fe0eee29dc7fdb1b50b54993bd              ERCC6   
8   01edd01a0414afd50a9964a16f911dc36cdad84f                NF1   
9   022396645475587dfd4d032e0673b7cec820369b              ZNRF3   
10  022b5518f8bff136a13d089b306bfbaf60d39ef3               EGFR   
11  026434a6ad26ace7ce8cb75ee8626448410e8c69                NF1   
12  0274d4abad6155af725ec04ee2ef16b042831471               DPYD   
13  02895720d3b2658d92af8292c6c14a0909e66c1c                NF1   
14  028e3ffad33f8c185374da203f8d2fcf5bd2e3c8               ABL1   
15  0301353b79d8f66acc7fcd10c197caa6bc894ca9                SMO   
16  033cec7fc45b36431c48064ecd8f7f4f00d8a0c3              PALB2   
17  037da9c2b79288b319df2aeb2985d9784c543eaa               JAK1   
18  037e9690938a8ee05b17f9c96df2401ebe15e654              ERBB2   
19  03b958d9f53bc50f7962c2c748d5f51a9f5f36ac               EGFR   
20  040d213cad0dd2a387d513f05cfdf12e1fc1807f              GNA11   
21  04389bc21fc5125398a124e5dcb35c5700e14495                MET   
22  0468f245bbf7fecfc82c6eae334b8b1f7397d3c8               NRAS   
23  04d05b28d6fdbb890a2ecc45bf5bc62050305632              ERBB2   
24  05387e19c94e509ba679c4f20ab5443606be53ca                KIT   
25  054dfbe2a634639e4d97e04aa617459c88acb4df               BRAF   
26  055fa2f24d33504086d6081b9706188decac33e1                ATR   
27  0576fff07da94fea5c370c68c4eb3a68d21e1bd5                NF2   
28  060c69c248b552be559953e1873d8cea3380ba26               ABL1   
29  068a71e2b184ce84f548e6c3f94b772f6bb7f25b                ALK   
30  06a0d71fc6dd86bde04d897e89be57a56404970f                NF2   
31  072ee1c871f79016315c983e8f4995490d5c08a3               HRAS   
32  073bc85bbf0e542a4b72627224abfcfb10f1017e              PBRM1   
33  075b37be7fcd3c9fe2c06e42a12a8fd5cca35b76               PAK1   
34  0773e656248c4358c42d5ea003e7f96e38990410               ABL1   
35  07b74b84f25244086ad9557129ed692e0acbe15b               TP53   
36  07b9dc6dd75c0ab2ecf3106cf59ffa0973fc81c7               NRAS   
37  083b88bebfe32a05e69f0db345c1fcef2d054206              CRLF2   
38  08a20b866cd7d4d258dc2421b99ce641c991b51b               BRAF   
39  08c5720b1197a1a56d3cb2d2665ad9c92b670fdb             PIK3CA   
40  08f40e1371fcba9ebde9568f2dd672903d80feed               BAP1   
41  08f91b7c77f82adc25a40960649c6561c1e3554b              BRCA1   
42  0903d548495248e842e18f233f5d4ad9b6285fed               CDK4   
43  09668ba4fdc2d4cf59015bef62efbad45bc1a137              AURKA   
44  0990ed8d8b5fae9ece1c6436437d209c53e8573f               ABL1   
45  09cbfdf92b11acffa3512e9e439c60c6e248fd82               ROS1   
46  09d2e4332f159c7aa111cd99433152d44037fc3a             DNMT3A   
47  09f5c420e6bbba58c4da18e1c73434f40bd653d7                NF2   
48  0a00bc45253da0f5eb1ff437dca2ed21d480eb95                NF1   
49  0a03f0cfe3d633f7be0752f72b7a6583d07c5d0b               KRAS   

   diseaseFromSourceMappedId       datasourceId        datatypeId  \
0                EFO_0003060  cancer_biomarkers  affected_pathway   
1                EFO_0005701  cancer_biomarkers  affected_pathway   
2                EFO_0000365  cancer_biomarkers  affected_pathway   
3                EFO_0001071  cancer_biomarkers  affected_pathway   
4                EFO_0001071  cancer_biomarkers  affected_pathway   
5                EFO_0000365  cancer_biomarkers  affected_pathway   
6                EFO_0005543  cancer_biomarkers  affected_pathway   
7                EFO_0001075  cancer_biomarkers  affected_pathway   
8  

In [34]:
# Compare score distributions
print(f"Associations score range: {a_df['score'].min():.2f} to {a_df['score'].max():.2f}")
print(f"Biomarkers score range: {b_df['score'].min():.2f} to {b_df['score'].max():.2f}")

# Are they similar distributions?
print("\nAssociations score stats:")
print(a_df['score'].describe())

print("\nBiomarkers score stats:")
print(b_df['score'].describe())

# How many unique target-disease pairs in each?
print("\nAssociations pairs:", a_df[['targetId', 'diseaseId']].drop_duplicates().shape[0])
print("Biomarkers pairs:", b_df[['targetId', 'diseaseId']].drop_duplicates().shape[0])

# How many overlap?
assoc_pairs = set(a_df[['targetId', 'diseaseId']].apply(tuple, axis=1))
biom_pairs = set(b_df[['targetId', 'diseaseId']].apply(tuple, axis=1))
print("Overlapping pairs:", len(assoc_pairs & biom_pairs))

Associations score range: 0.00 to 0.91
Biomarkers score range: 1.00 to 1.00

Associations score stats:
count    4.492971e+06
mean     6.339187e-02
std      1.085055e-01
min      7.902492e-04
25%      3.695799e-03
50%      2.197744e-02
75%      5.335624e-02
max      9.133829e-01
Name: score, dtype: float64

Biomarkers score stats:
count    1301.0
mean        1.0
std         0.0
min         1.0
25%         1.0
50%         1.0
75%         1.0
max         1.0
Name: score, dtype: float64

Associations pairs: 4492971
Biomarkers pairs: 469
Overlapping pairs: 469


## Score Column Analysis

The `score` columns in both datasets are **semantically distinct** and cannot be directly compared:

**Associations dataset:**
- Score represents the aggregated strength of evidence for a direct target-disease association
- Calculated across all data sources (GWAS, literature, animal models, etc.)
- Range: 0.01 to 0.83 in our data
- Interpretation: Higher score = stronger overall evidence linking target to disease

**Biomarkers dataset:**
- Score is always 1.0 (binary indicator that biomarker evidence exists)
- Does not provide a meaningful confidence metric
- The actual evidence quality is captured in the `confidence` column instead (values: "case report", "late trials", "pre-clinical", etc.)
- Represents strength of evidence for a specific biomarker observation, not the overall target-disease relationship

## Dataset Relationship

Key findings from overlap analysis:
- **Associations:** 4,492,971 unique target-disease pairs
- **Biomarkers:** 469 unique target-disease pairs
- **Overlap:** All 469 biomarker pairs exist in the associations dataset (100% overlap)

This confirms that the biomarkers dataset is a **focused subset** of associations, providing drug-biomarker enrichment data for cancer-specific target-disease relationships. The biomarkers data represents detailed evidence for a small fraction (~0.01%) of all known target-disease associations.

In [35]:
# Can one target-disease pair have multiple biomarker records?
biomarker_counts = b_df.groupby(['targetId', 'diseaseId']).size()
print("Biomarker records per target-disease pair:")
print(biomarker_counts.describe())
print("\nMax biomarkers for one pair:", biomarker_counts.max())

Biomarker records per target-disease pair:
count    469.000000
mean       2.773987
std        4.208773
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       42.000000
dtype: float64

Max biomarkers for one pair: 42


## Cardinality Analysis

The relationship between associations and biomarkers is **one-to-many**:
- Most target-disease pairs (50%) have just 1 biomarker record
- Mean: 2.77 biomarker records per pair
- Maximum: 42 records for a single pair
- Distribution is heavily right-skewed

**Why one-to-many?**
A single target-disease association can have multiple biomarker evidence items, each documenting different drugs, genetic variants, clinical contexts, or confidence levels.

**Example:** BRCA1-breast cancer (one association) might have separate biomarker records for trastuzumab response, olaparib efficacy, and various clinical trial findings.

**Implication for data model:** Associations and biomarkers should be kept in **separate, related tables** rather than flattened into one denormalized table.